sys/exec.c

```diff
@@ -179,10 +179,13 @@ int do_exec(const exec_args_t *args) {
         vm_addr_t end = roundup(ph->p_vaddr + ph->p_memsz, PAGESIZE);
         /* TODO: What if segments overlap? */
         /* Temporarily permissive protection. */
-        vm_map_entry_t *segment = vm_map_add_entry(
-          vmap, start, end, VM_PROT_READ | VM_PROT_WRITE | VM_PROT_EXEC);
-        /* Allocate pages backing this segment. */
-        segment->object = default_pager->pgr_alloc();
+        vm_map_entry_t *segment;
+        WITH_MTX_LOCK (&vmap->mtx) {
```

Comment:
Looking at the following code I'm almost convinced that `vm_map` interface is badly designed. IMHO forcing user to lock on `vm_map`'s is not necessary here. Ideally this code should look as follows:
```c
vm_object_t *obj = vm_object_alloc(VM_OBJ_DEFAULT);
vm_map_entry_t *entry = vm_map_insert(vmap, vobj, start, end,
                                      VM_PROT_READ | VM_PROT_WRITE | VM_PROT_EXEC);
```
With following procedure signatures:
```c
vm_object_t *vm_object_alloc(vm_object_type_t);
int vm_map_insert(vm_map_t *map, vm_object_t *object, 
                  vm_offset_t start, vm_offset_t end, vm_prot_t prot);
```

What do you think?

Reply:
I agree, the proposed interface is much like the one in FreeBSD. Should this change go in a separate PR?

Reply:
Yes, please do so.

Reply:
Please introduce both procedures in this PR.